### ISSUE TITLE
Cross jobs failing on Hydra

### ISSUE THREAD

**Issue by matthewbauer** [NixOS/nixpkgs#42407]
This is a meta-issue for tracking failing cross compilation jobs on Hydra. Please help us out and investigate causes and find potential solutions! We want to 18.09 to have great cross compilation support.

- [x] [aarch64-musl.nixUnstable.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/aarch64-musl.nixUnstable.x86_64-linux)
- [ ] [android64.nix.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/android64.nix.x86_64-linux)
- [ ] [android64.nixUnstable.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/android64.nixUnstable.x86_64-linux)
- [ ] [crossMingw32.haskell.packages.ghc822.hello.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/crossMingw32.haskell.packages.ghc822.hello.x86_64-linux)
- [ ] [crossMingw32.haskell.packages.ghcHEAD.hello.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/crossMingw32.haskell.packages.ghcHEAD.hello.x86_64-linux)
- [x] [crossMingw32.libunistring.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/crossMingw32.libunistring.x86_64-linux)
- [ ] [crossMingw32.windows.wxMSW.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/crossMingw32.windows.wxMSW.x86_64-linux)
- [ ] [crossMingwW64.haskell.packages.ghc822.hello.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/crossMingwW64.haskell.packages.ghc822.hello.x86_64-linux)
- [ ] [crossMingwW64.haskell.packages.ghcHEAD.hello.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/crossMingwW64.haskell.packages.ghcHEAD.hello.x86_64-linux)
- [x] [crossMingwW64.libunistring.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/crossMingwW64.libunistring.x86_64-linux)
- [ ] [crossMingwW64.windows.wxMSW.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/crossMingwW64.windows.wxMSW.x86_64-linux)
- [x] [rpi-musl.haskell.packages.ghcHEAD.hello.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/rpi-musl.haskell.packages.ghcHEAD.hello.x86_64-linux)
- [x] [rpi-musl.nixUnstable.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/rpi-musl.nixUnstable.x86_64-linux)
- [x] [rpi.haskell.packages.ghcHEAD.hello.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/rpi.haskell.packages.ghcHEAD.hello.x86_64-linux)
- [x] [rpi.nixUnstable.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/rpi.nixUnstable.x86_64-linux)
- [x] [x86_64-musl.haskell.packages.ghc822.hello.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/x86_64-musl.haskell.packages.ghc822.hello.x86_64-linux)
- [x] [x86_64-musl.haskell.packages.ghcHEAD.hello.x86_64-linux](https://hydra.nixos.org/job/nixpkgs/cross-trunk/x86_64-musl.haskell.packages.ghcHEAD.hello.x86_64-linux)

The file that lists these is pkgs/top-level/release-cross.nix

**Comment by dtzWill** [comment]
The nixUnstable fails on armv7l (muslpi) and aarch64:
* same error building against glibc vs musl
* I guess there's no support for fcontext (boost context asm bits) on these platforms? Or maybe there is but it fails to recognize our triples or something?
* We may be able to build with BOOST_USE_UCONTEXT as a workaround, although that's known slower (and perhaps might have other implications, I'm not sure).  I don't know if there's a good way for these flags to be neatly propagated-- so downstream users automagically use the same flags used to build boost...?

**Comment by dtzWill** [comment]
update: for some reason the "toolset" argument isn't evaluating the way we expect it to-- it defaults to "if <usingclang> then "clang" else <if cross and using gcc> then "gcc-cross" else null", but it seems to be null anyway--in fact inserting a trace shows it doesn't get evaluated anyway?

Perhaps due to the way things are chained from the other files into `generic.nix`?

No idea what happens if we set "toolset=gcc-cross" as this looks like we intend to do, but seems like a good place to start investigating.... :)

**Comment by dtzWill** [comment]
ghc failures seem to be due to wanting to execute `opt` which doesn't work on cross since llvm is a `propagatedBuildInput` -- we'll need to plumb in the path to a build-native `opt` or so?

**Comment by matthewbauer** [comment]
Also @Ericson2314 do you think we could add at least 1 cross compiled NixOS to release-cross.nix? It would add a little bit of cost but it is the ultimate test of whether our cross toolchain actually works.

**Comment by matthewbauer** [comment]
Does anyone know what ```undefined reference to `__stack_chk_guard'``` means? Could this be related to "stackprotector" hardening? I wonder if that needs to be disabled for mingw.

**Comment by matthewbauer** [comment]
Is stack protector broken with musl?

**Comment by matthewbauer** [comment]
Is stack protector just broken in Musl?

**Comment by Ericson2314** [comment]
@matthewbauer cross nixos sounds good.

**Comment by Ericson2314** [comment]
@dtzWill I thought GHC used strict deps so Wich things wouldn't appear on the PATH.

**Comment by dtzWill** [comment]
On 32bit IIRC it's a mild SNAFU re:ssp -- solved slightly differently on different musl-based distributions, I forget the details. I've tried various approaches to addressing this in branches on my fork, but think was still testing what the best solution was. And since we dropped 32bit support in theory, I didn't finish looking into this.

You're only seeing it on i686 right?

----

GHC should work now -- ghc822 anyway, as of PR merged yesterday. :D

**Comment by dtzWill** [comment]
If we're looking to get all of these passing (great goal!) it may make sense to entirely disable the i686-musl bits for now.

**Comment by angerman** [comment]
I don't think we can (or should) cross compile coreutils with mingw. As far as I know only cygwin is supported by coreutils.

**Comment by matthewbauer** [comment]
@angerman I was under the impression that it would work. At least mingw published binaries for it at one point:

https://sourceforge.net/projects/mingw/files/MSYS/Base/coreutils/

**Comment by angerman** [comment]
Note that those are some ancient version 5 (current coreutils is 8) packages. I haven’t tried them, but cross compiling coreutils with mingw right now results in:

```
configure: error: could not determine how to read list of mounted file systems
```

See also https://stackoverflow.com/a/8355683

**Comment by matthewbauer** [comment]
Ok thanks for the info. I have removed coreutils mingw here but it is still in release-cross.nix so should probably be removed.

Does anyone know what the issue is with LLVM failures:

https://hydra.nixos.org/build/78560084

?

**Comment by Ericson2314** [comment]
Those LLVMs shouldn't even be cross-specific.

**Comment by matthewbauer** [comment]
Fixed the LLVM issue in f3fc04bfd2cc3889236d7b31e8555a319f0a74d5.

**Comment by matthewbauer** [comment]
Closing for now. We’ve added a bunch of new jobs and fixed most of the original from this issue.
